File: csharp-selenium-nunit-reqnroll/SauceDemo.UiTests/StepDefinitions/CartSteps.cs
using NUnit.Framework;
using OpenQA.Selenium;
using Reqnroll;
using SauceDemo.UiTests.Pages;

namespace SauceDemo.UiTests.StepDefinitions;

[Binding]
public sealed class CartSteps
{
    private readonly IWebDriver _driver;

    public CartSteps(IWebDriver driver)
    {
        _driver = driver;
    }

    [Then(@"the cart badge should show (\d+)")]
    public void ThenTheCartBadgeShouldShow(int expectedCount) =>
        Assert.That(
            () => new InventoryPage(_driver).GetCartCount(),
            Is.EqualTo(expectedCount).After(5000, 250));

    [When(@"I remove ""(.*)"" from the cart")]
    public void WhenIRemoveFromTheCart(string productName) => new InventoryPage(_driver).RemoveFromCart(productName);

    [When(@"I start checkout")]
    public void WhenIStartCheckout() => new CartPage(_driver).StartCheckout();

    [Then(@"the cart should list the following products:")]
    public void ThenTheCartShouldListTheFollowingProducts(DataTable table)
    {
        var expected = table.Rows.Select(r => r["ProductName"]).ToList();
        Assert.That(
            () => new CartPage(_driver).ListItemNames(),
            Is.EquivalentTo(expected).After(5000, 250));
    }
}

File: ts-playwright-cucumber/src/steps/cartSteps.ts
import { DataTable, Then, When } from '@cucumber/cucumber';
import { expect } from 'chai';
import { CartPage } from '../pages/cartPage.ts';
import { InventoryPage } from '../pages/inventoryPage.ts';
import { retryUntil } from '../support/retryUntil.ts';
import type { SauceDemoWorld } from '../world.ts';

Then('the cart badge should show {int}', async function (this: SauceDemoWorld, expectedCount: number) {
  const inventoryPage = new InventoryPage(this.page);
  const count = await retryUntil(
    () => inventoryPage.getCartCount(),
    (c) => c === expectedCount
  );
  expect(count).to.equal(expectedCount);
});

When('I remove {string} from the cart', async function (this: SauceDemoWorld, productName: string) {
  await new InventoryPage(this.page).removeFromCart(productName);
});

When('I start checkout', async function (this: SauceDemoWorld) {
  await new CartPage(this.page).startCheckout();
});

Then('the cart should list the following products:', async function (this: SauceDemoWorld, table: DataTable) {
  const expected = table.hashes().map((row) => row.ProductName ?? '');
  const cartPage = new CartPage(this.page);
  const names = await retryUntil(
    () => cartPage.listItemNames(),
    (actual) => actual.length === expected.length && actual.every((name) => expected.includes(name))
  );
  expect(names).to.have.members(expected);
});
